core/packages/contracts/contracts/SovereignTreasury.sol

```diff
@@ -0,0 +1,49 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.9;
+
+import "./Vault.sol";
+
+// TODO: transfer ownership from deployer to inbound channel
+contract SovereignTreasury is Ownable {
+    Vault public vault;
+
+    constructor(Vault _vault) {
+        vault = _vault;
+    }
+
+    // Handle a message from the bridge.
+    function handle(bytes32 _sovereignID, bytes calldata _message) external onlyOwner {
+        Message memory message = abi.decode(_message, (Message));
+
+        if (message.action == Action.Withdraw) {
+            WithdrawPayload memory payload = abi.decode(message.payload, (WithdrawPayload));
+
+            transfer(_sovereignID, payload.recipient, payload.amount);
+        }
```

Comment:
We should revert if the message is not an action we know about.

Reply:
Agree.

Should raise a custom error like `UnknownMessage`: https://blog.soliditylang.org/2021/04/21/custom-errors/

Reply:
Won't `abi.decode` have to revert if the action isn't recognised? Else it has to choose an action for `message` to be defined.

Makes sense to have a better error message for it 👌

Reply:
I have tested in Remix VM and foundry VM and enums are validated when using `abi.decode`. So the code that reverts with that custom error will be unreachable.